design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,344 @@
+
+Supposing we used an inline handlebar template syntax, the `deployment` field could contain:
+
+```yaml
+containers:
+- name: redis-controller
+  image: redis/redis-crossplane-controller:{{.spec.version}}
+  imagePullPolicy: Always
+```
+
+_In practice, `version` would most likely be static, pinned to a specific version of a Stack_.
+
+In this example, the `.spec.version` value would be taken from the body of each CR this Stack manages. That value would be applied to the template and rendered as a new or updated `Deployment` resource.
+
+Should the `controller.deployment` value be treated as a template? This could have unintended consequences for existing Stack deployments, especially if their definitions includes pieces that match the template format.
+
+To avoid disturbing existing `controller.deployment` values, let's use a new template aware `templates` spec field. We'll take that approach in the following sections.
+
+### No Custom Controller Needed
+
+Existing Stacks must specify a `Deployment`. These are typically written in Go and generally require understanding some combination of client-go, controller-runtime, crossplane-runtime, and Kubebuilder.
+
+For some applications, the intended work of the controller could be simple:
+
+* For every new resource, create a set of Kubernetes resources
+* Use the spec fields of that resource to affect the Kubernetes resources
+* Upon changes to these spec fields, update the Kubernetes resources
+
+This is where the value of Template Stacks starts to materialize.
+
+```yaml
+kind: Stack
+metadata:
+  name: RedisStack
+spec:
+  templates:
+    redis.example.org/v1:
+      deployment: |
+          …
+          containers:
+          - name: redis-controller
+            image: example/redis-controller:{{.spec.redisVersion}}
+```
+
+*In this example, templates are nested within a map keyed by the GVR and a template grouping name. This will be explained in the [One Pager for Template Stack Fields](one-pager-template-stack-fields).*
+
+### Composition
+
+#### Vertical Composition
+
+Crossplane Stacks manage multiple CRDs. Template Stacks should be no different.
+
+Template definitions will need to express the template to render for each CRD that is managed. The GVR can be used as a map key to denote the CRD template being described.
+
+Each template needed to render a GVR should be provided (demonstrated below as a map), additionally a name that represents that intent of that template should distinguish each template (also demonstrated as a map). This name can then be reused to reference a template (see [Horizontal Composition](#Horizontal-Composition)).
+
+```yaml
+kind: Stack
+metadata:
+  name: CachingWebServiceStack
+spec:
+  templates:
+    redis.example.org/v1:
+      deployment: |
+        kind: Deployment
+        …
+          containers:
+          - name: redis-controller
+            image: example/redis-controller:{{.spec.redisVersion}}
```

Comment:
circular dependencies need to handled/disallowed.

Reply:
I should point out in the design that the controller doesn't need to be concerned with this (although the template developer does).

The rendering process of the Template Stack Manager will use a single (cached) pass at applying templates.

* All of the current resources are fetched
* The resources are combined to form a variables to pass to the template
* Template substitutions are made for each resource
* All of the templated resources are applied to the API
  Including a pass at deleting resources that whose templates have become empty

I do not believe the template rendering process can get caught in a circular dependency loop here. However, the Template Stack creator could improperly craft their templates in a way that they would never resolve.  This could be left as a `crossplane-cli` concern.

Similarly, if a template contained invalid resource yaml, or contained fields and values that were not valid for the resource being generated, the Template Stack Manager will need to a defined means of surfacing that error condition (perhaps as a template variable or function, but also through logging).

I intended to include some of this thinking in the design - thanks for calling out the omissions!

Reply:
This will be found in the `Life Cycle and Updates` section which was woefully bare.